src/actions/sources/loadSourceText.js

```diff
@@ -25,17 +28,35 @@ async function loadSource(source: Source, { sourceMaps, client }) {
   };
 }
 
+function defer() {
+  let resolve = () => {};
+  let reject = () => {};
+  const promise = new Promise((_res, _rej) => {
+    resolve = _res;
+    reject = _rej;
+  });
+
+  return { resolve, reject, promise };
+}
+
 /**
  * @memberof actions/sources
  * @static
  */
 export function loadSourceText(source: Source) {
   return async ({ dispatch, getState, client, sourceMaps }: ThunkArgs) => {
+    const deferred = defer();
+
     // Fetch the source text only once.
-    if (source.text) {
+    if (isLoaded(source)) {
       return Promise.resolve(source);
     }
 
+    if (isLoading(source) || requests.has(source.id)) {
+      return requests.get(source.id);
+    }
+
+    requests.set(source.id, deferred.promise);
```

Comment:
we set the promise here because we only want to resolve it after symbols and other async work is done